cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
+                String selectedVoltageLevelId = representativeVoltageLevelId(vlAds);
+                for (String voltageLevelId : vlAds) {
+                    if (!voltageLevelId.equals(selectedVoltageLevelId)) {
+                        voltageLevelMapping.put(voltageLevelId, selectedVoltageLevelId);
                     }
-                });
-        if (!mapping.isEmpty()) {
+                }
+            }
+        });
+        if (!voltageLevelMapping.isEmpty()) {
+            LOG.warn("VoltageLevel id mapping needed for {} voltageLevels: {}",
+                voltageLevelMapping.size(), voltageLevelMapping);
+        }
+    }
+
+    private void buildSubstation(Map<String, List<String>> substationAdjacency) {
+        Set<String> visitedSubstations = new HashSet<>();
+        substationAdjacency.keySet().forEach(sub -> {
+            if (!visitedSubstations.contains(sub)) {
+                ArrayList<String> subAds = allConnected(substationAdjacency, visitedSubstations, sub);
+
+                String selectedSubstationId = representativeSubstationId(subAds);
+                for (String substationId : subAds) {
+                    if (!substationId.equals(selectedSubstationId)) {
+                        substationMapping.put(substationId, selectedSubstationId);
+                    }
+                }
+            }
+        });
+        if (!substationMapping.isEmpty()) {
             LOG.warn("Substation id mapping needed for {} substations: {}",
-                    mapping.size(), mapping);
+                    substationMapping.size(), substationMapping);
         }
     }
 
+    // Given an id (substation / voltageLevel) returns all connected ids (substations / voltageLevels)
+    // Two ids are connected if they are adjacent in the adjacency Map
+    private static ArrayList<String> allConnected(Map<String, List<String>> adjacency, Set<String> visited, String id) {
+        ArrayList<String> allConnected = new ArrayList<>();
+
+        // Insert id in the allConnected list and record it as visited
+        allConnected.add(id);
+        visited.add(id);
+
+        // Expand, adding in each step all non-visited adjacent ids"
+        int k = 0;
+        while (k < allConnected.size()) {
+            String vl0 = allConnected.get(k);
+            if (adjacency.containsKey(vl0)) {
+                adjacency.get(vl0).forEach(ad -> {
+                    if (visited.contains(ad)) {
+                        return;
+                    }
+                    allConnected.add(ad);
+                    visited.add(ad);
+                });
+            }
+            k++;
+        }
+        return allConnected;
+    }
+
+    private static String representativeVoltageLevelId(Collection<String> voltageLevelIds) {
+        return voltageLevelIds.stream()
+                .sorted()
+                .findFirst()
+                .orElseThrow(() -> new PowsyblException("Unexpected: voltageLevelIds list is empty"));
```

Comment:
You should throw an AssertionError if this list is expected to never be empty

Reply:
Done